pytket/extensions/cutensornet/mps/mps.py

```diff
@@ -0,0 +1,735 @@
+
+        if len(positions) == 1:
+            self._apply_1q_gate(positions[0], gate.op)
+            # NOTE: if the tensor was in canonical form, it remains being so,
+            #   since it is guaranteed that the gate is unitary.
+
+        elif len(positions) == 2:
+            dist = positions[1] - positions[0]
+            # We explicitly allow both dist==1 or dist==-1 so that non-symmetric
+            # gates such as CX can use the same Op for the two ways it can be in.
+            if dist not in [1, -1]:
+                raise RuntimeError(
+                    "Gates must be applied to adjacent qubits! "
+                    + f"This is not satisfied by {gate}."
+                )
+            self._apply_2q_gate(positions, gate.op)
+            # The tensors will in general no longer be in canonical form.
+            self.tensors[positions[0]].canonical_form = None
+            self.tensors[positions[1]].canonical_form = None
+
+        else:
+            raise RuntimeError(
+                "Gates must act on only 1 or 2 qubits! "
+                + f"This is not satisfied by {gate}."
+            )
+        return self
+
+    def canonicalise(self, l_pos: int, r_pos: int) -> None:
+        """Apply the necessary gauge transformations so that all MPS tensors
+        to the left of position ``l_pos`` are in left orthogonal form and
+        all MPS tensors to the right of ``r_pos`` in right orthogonal form.
+
+        Args:
+            l_pos: The position of the leftmost tensor that is not to be
+                canonicalised.
+            r_pos: The position of the rightmost tensor that is not to be
+                canonicalised.
+        """
+        for pos in range(l_pos):
+            self.canonicalise_tensor(pos, form=DirectionMPS.LEFT)
+        for pos in reversed(range(r_pos + 1, len(self))):
+            self.canonicalise_tensor(pos, form=DirectionMPS.RIGHT)
+
+    def canonicalise_tensor(self, pos: int, form: DirectionMPS) -> None:
+        """Apply the necessary gauge transformations so that the tensor at
+        position ``pos`` in the MPS has is in the orthogonal form dictated by
+        ``form``.
+
+        Args:
+            position: The position of the tensor to canonicalise.
+            form: LEFT form means that its conjugate transpose is its inverse if
+                connected to its left bond and physical bond. Similarly for RIGHT.
+        """
+        if form == self.tensors[pos].canonical_form:
+            # Tensor already in canonical form, nothing needs to be done
+            return None
+
+        if self._libhandle is None:
+            raise RuntimeError(
+                "Must be called inside a with mps.init_cutensornet() block."
+            )
+
+        if form == DirectionMPS.LEFT:
+            next_pos = pos + 1
+            gauge_bond = pos + 1
+            gauge_T_index = 0
+            gauge_Q_index = -2
+        elif form == DirectionMPS.RIGHT:

```

Comment:
Since I'm using an enum type, the type checker would complain if you pass it something that isn't a value of  the enum... but still I think you are right, since users are likely not to use mypy.

Reply:
Yeah, it won't be picked up at runtime. That said, we don't necessarily need to throw in case of any input inconsistencies. API is clear.